### PR TITLE
Basic implementation of completion for Modelica code

### PR DESCRIPTION
**Pull request by atrosinenko**
Complete everything against the latest successfully saved version of the top-level class opened in the current editor.

Partially fixes [this ticket](https://trac.openmodelica.org/OpenModelica/ticket/3847).

This should already be better than nothing. On the other hand, here are things that probably should be done as well. Should they go to this single PR?

- [x] Basic implementation
- [x] Optimization
- [x] Add an icon for new completion type
- [ ] Add **different** icons for different symbol types (how many?)
- [ ] Respect **class** under cursor (consider nested classes in single file)
- [ ] Check multiple inheritance traversal order

### CODE REVIEW

**Comment by atrosinenko** [comment]
Added an icon made of existing completerType.svg. Hope it isn't against CLA.

**Comment by adeas31** [comment]
Performance has the high priority here. I believe we should stay inside Qt data structures as much as possible and should avoid calls to OMC. All the calls you are doing to OMC are already done and should be avoided. For example, right now if you create a new model and then switch to text view and just press ctrl+space, as a result of this action you are sending 10 commands to OMC whose response time depends very much on the model. I suggest the following,

* Fetch the information of loaded classes from the Libraries Browser.
* Fetch the information of the inherited classes and components from the `ModelWidget` class.

**Comment by atrosinenko** [comment]
Now it should not call backend without necessity. But now code review is even more required.

**Comment by atrosinenko** [comment]
Quick fix: collect completion from outer packages. Now there is a false positive `All`, but completion anyways is not precise now...

**Comment by atrosinenko** [comment]
Autocompletion is a very broad topic, so what is expected from me to be done in this PR and what can be left as is / for the next PRs?

**Comment by adeas31** [comment]
Unfortunately I am busy with other things and don't have enough time to review your changes. I don't promise anything but will try to review this or next week.

**Comment by atrosinenko** [comment]
I just asked in case there is something implicitly expected from me that I don't realize. If you are busy then please excuse me for disturbance, my previous post was not about "review this ASAP". :)

**Comment by atrosinenko** [comment]
Is there any updates on this PR? Is it postponed till 1.13 release?

**Comment by adeas31** [comment]
Yes. I think we should not push this in until the release.

**Comment by atrosinenko** [comment]
OK. Thank you.

**Comment by atrosinenko** [comment]
Ping :)

File: OMEdit/OMEditGUI/Modeling/LibraryTreeWidget.cpp [hunk 1]
```diff
@@ -605,6 +605,85 @@ void LibraryTreeItem::removeInheritedClasses()
...
+}
+
+void LibraryTreeItem::tryToComplete(QSet<QString> &result, const QString &lastPart)
+{
+  QList<LibraryTreeItem*> baseClasses = getInheritedClassesDeepList();
+
+  for (int bc = 0; bc < baseClasses.size(); ++bc) {
+    QList<LibraryTreeItem*> classes = baseClasses[bc]->childrenItems();
+    for (int i = 0; i < classes.size(); ++i) {
+      if (classes[i]->getName().startsWith(lastPart))
```
Comment (adeas31) [inline OMEdit/OMEditGUI/Modeling/LibraryTreeWidget.cpp]: Yes `All` is a special `LibraryTreeItem` that is used for search functionality but is hidden and not shown to the user. You need to filter it here.

`if (classes[i]->getName().startsWith(lastPart) && classes[i]->getNameStructure().compare("OMEdit.Search.Feature") != 0)`

File: OMEdit/OMEditGUI/Modeling/LibraryTreeWidget.cpp [hunk 2]
```diff
@@ -605,6 +605,85 @@ void LibraryTreeItem::removeInheritedClasses()
...
+        result.insert(classes[i]->getName());
+    }
+
+    const QList<ComponentInfo*> &components = baseClasses[bc]->getComponentsList();
+    for (int i = 0; i < components.size(); ++i) {
+      if (components[i]->getName().startsWith(lastPart))
+        result.insert(components[i]->getName());
+    }
```
Comment (adeas31) [inline OMEdit/OMEditGUI/Modeling/LibraryTreeWidget.cpp:697]: The root `LibraryTreeItem` is just a container. It is not a valid Modelica class so asking for its components will lead to an error. You need to check for it here,

```
if (!baseClasses[bc]->isRootItem()) {
      const QList<ComponentInfo*> &components = baseClasses[bc]->getComponentsList();
      for (int i = 0; i < components.size(); ++i) {
        if (components[i]->getName().startsWith(lastPart))
          result.insert(components[i]->getName());
      }
    }
```

Note that OMEdit can also load non-Modelica items like OMSimulator models, composite models. text files etc. So you need to check for it as well i.e., add the following to the above if condition `baseClasses[bc]->getComponentType() == LibraryTreeItem::Modelica`

**Review by adeas31 — CHANGES_REQUESTED** [review]
LGTM. What are the plans for the remaining items in the PR?

> Add different icons for different symbol types (how many?)

I think we should just consider classes and components. So 2 icons one for classes and one for components should be enough. If easier maybe we can differentiate between inherited classes. However, I believe the icon is not so important. I think we should improve the details of the tooltip. In the tooltip we should show more information about the selected completer item.

There are few minor things that I have noticed while testing,

- The first popup item should be selected so I can use it by pressing enter. Right now I have to press down arrow and then enter.
- We should show the popup after the `.` is pressed. For example, user types Modelica and then `.` after that show the popup. Right now I have to press ctrl+space.

**Comment by atrosinenko** [comment]
Remaining items are more for reviewers: these are some ideas on what should be done right now and what can be postponed to subsequent PRs. Personally I think it is worth to have some "beta that is better than nothing" version of completion right now and improve it later instead of trying to bring in some ideal completion functionality in one merge (that is, postpone remaining items to future PRs).

Thank you for your review comments, I hope I will fix them soon.

**Comment by atrosinenko** [comment]
Amended my PR as suggested and hopefully fixed one of "minor things" (popping up of completion on `.`). I have not managed to implement selection of the first completion item, though.

**Comment by casella** [comment]
@AndreaBartolini, you may be interested in watching this progress

**Comment by adeas31** [comment]
> Amended my PR as suggested and hopefully fixed one of "minor things" (popping up of completion on `.`).

Thanks for the fixes.

> I have not managed to implement selection of the first completion item, though.

Just wondering if you are still working on it or should we merge this PR then?

**Comment by atrosinenko** [comment]
Ultimately, the first item becomes auto-selected... I still cannot understand why current item sometimes becomes de-synchronized with the visually selected one. But even if it is regression of this PR (I cannot quickly prove / disprove it because of quite restricted hardcoded completion item set in the master branch), this looks not too significant comparing to the possibility to at least see (and select with partial visual feedback, in fact) the next word to enter.

> Just wondering if you are still working on it or should we merge this PR then?

I would prefer to merge at least something, and then improve later.

**Comment by adeas31** [comment]
Tested a bit more and it seems to work well.
Can you squash the commits? Makes it easier if we want to port it to 1.13 release.

**Comment by atrosinenko** [comment]
I can squash the commits 2-3 hours later. Maybe GitHub can squash the commits automatically if properly ask it?

**Comment by adeas31** [comment]
We have that option disabled for the PRs. We had have issues with it in the past. It works better with the manual squash and then rebase and merge of PR.

**Comment by atrosinenko** [comment]
OK, no problem at all. I hope to merge it about 2 hours later.

**Comment by atrosinenko** [comment]
Pushed the result of `git rebase -i master` with squashing.

**Comment by casella** [comment]
@atrosinenko, thanks a lot for your contribution, I just tried it and it's really nice.

Can you please register on trac.openmodelica.org, so I can assign ticket [#3847](https://trac.openmodelica.org/OpenModelica/ticket/3847) to you?

I think the most important missing feature now is to take care of inheritance. For example, if you type Modelica.Media.Interfaces.PartialTwoPhaseMedium. (which is something I do often, BTW), only the classes defined or re-declared there show up. In fact, there are many others defined or re-defined in the ancestor classes Modelica.Media.Interfaces.PartialPureSubstance and Modelica.Media.Interfaces.PartialMedium. 

If the drop-down list showed up also the classes defined or re-declared within all the ancestor classes, it would be a real bonus, because it's quite hard, particularly for beginner users of Modelica.Media, to figure out what is defined and where.

If you need to fetch this information from the API, you can ask @adrpo and @perost how to do it.

Thanks!

**Comment by atrosinenko** [comment]
Thanks @adeas31 for review comments!

> Can you please register on trac.openmodelica.org, so I can assign ticket #3847 to you?

Looks like users are manually registered by @adrpo according to [this page](https://openmodelica.org/developersresources/source-code).

**Comment by adrpo** [comment]
@atrosinenko: Yes. I will make you a user in Trac with the same username as here.

**Comment by adrpo** [comment]
@atrosinenko: user created, check your email.

**Comment by atrosinenko** [comment]
@adrpo: Received, thanks!

**Pull request merged**